File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/browsing/library/access/FakeLibraryProvider.java
package com.lasthopesoftware.bluewater.client.browsing.library.access;

import com.annimon.stream.Stream;
import com.lasthopesoftware.bluewater.client.browsing.library.repository.Library;
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId;
import com.namehillsoftware.handoff.promises.Promise;

import org.jetbrains.annotations.NotNull;

import java.util.Collection;

public class FakeLibraryProvider implements ILibraryProvider {
	private final Library[] libraries;

	public FakeLibraryProvider(Library... libraries) {
		this.libraries = libraries;
	}

	@NotNull
	@Override
	public Promise<Library> getLibrary(@NotNull LibraryId libraryId) {
		return new Promise<>(Stream.of(libraries).filter(l -> l.getLibraryId().equals(libraryId)).findFirst().get());
	}

	@NotNull
	@Override
	public Promise<Collection<Library>> getAllLibraries() {
		return new Promise<>(Stream.of(libraries).toList());
	}
}

File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/browsing/library/access/FakeLibraryProvider.kt
package com.lasthopesoftware.bluewater.client.browsing.library.access

import com.lasthopesoftware.bluewater.client.browsing.library.repository.Library
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId
import com.lasthopesoftware.bluewater.shared.promises.extensions.toPromise
import com.namehillsoftware.handoff.promises.Promise

class FakeLibraryProvider(vararg libraries: Library) : ILibraryProvider, ILibraryStorage {
	private val libraries = libraries.associateBy { l -> l.id }.toMutableMap()

    override fun promiseLibrary(libraryId: LibraryId): Promise<Library?> = Promise(libraries[libraryId.id])

    override val allLibraries: Promise<Collection<Library>>
        get() = Promise(libraries.values)

	override fun saveLibrary(library: Library): Promise<Library> =
		library.copy().also { libraries[it.id] = it }.toPromise()

	override fun removeLibrary(library: Library): Promise<Unit> {
		libraries.remove(library.id)
		return Unit.toPromise()
	}
}
